sympy/solvers/solveset.py

```diff
@@ -1754,6 +1758,57 @@ def _is_logarithmic(f, symbol):
     return rv
 
 
+def _is_lambert(f):
+    r"""
+    Return True if the equation is of Lambert type, else False.
+    Equations containing `Pow`, `log` or `exp` terms are currently
+    treated as Lambert types.
+    """
+    return any(isinstance(arg, (Pow, exp, log)) for arg in _term_factors(f))
+
+
+def _compute_lambert_solutions(lhs, rhs, symbol):
+    r"""
+    Computes the equations whose solutions will be given in terms
+    of the list of LambertW function. Returns `None` if it fails doing so.
+
+    Example
+    =======
+
+    >>> from sympy.solvers.solveset import _compute_lambert_solutions
+    >>> from sympy import symbols, exp
+    >>> x = symbols('x')
+    >>> _compute_lambert_solutions(x*exp(x), 1, x)
+    [LambertW(1)]
+
+    """
+    try:
+        poly = lhs.as_poly()
+        gens = _filtered_gens(poly, symbol)
+        return _solve_lambert(lhs - rhs, symbol, gens)
+    except NotImplementedError:
+        pass
+
+
+def _solve_as_lambert(lhs, rhs, symbol, domain):
+    r"""
+    Helper solver to handle equations having Lambert solutions.
+    Initally tries to solve equation directly.
+
+    Notes
+    =====
+    This function needs more improvements.
+
+    """
+    result = ConditionSet(symbol, Eq(lhs - rhs, 0), domain)
+
+    soln = _compute_lambert_solutions(lhs, rhs, symbol)
+    if soln:
+        result = FiniteSet(*soln)
+
+    return result
```

Comment:
This could be clearer like:
```
    soln = _compute_lambert_solutions(lhs, rhs, symbol)
    if soln:
        return FiniteSet(*soln)
    else:
        return ConditionSet(symbol, Eq(lhs - rhs, 0), domain)
```